packages/gems/js/lib/js.rb

```diff
@@ -146,9 +158,88 @@ def new(*args)
   #
   #   JS.eval("return [1, 2, 3]").to_a.map(&:to_i)    # => [1, 2, 3]
   #   JS.global[:document].querySelectorAll("p").to_a # => [[object HTMLParagraphElement], ...
-  def to_a
+  def to_a(convertTypes: true)
     as_array = JS.global[:Array].from(self)
-    Array.new(as_array[:length].to_i) { as_array[_1] }
+    Array.new(as_array[:length].to_i) do
+      item = as_array[_1]
+      convertTypes and item.respond_to?(:to_rb) ? item.to_rb : item
+    end
+  end
+
+  # Try to convert JS Objects into Ruby Objects
+  # Todo: make a list of Types that need to remain JS::Objects (array methods?)
+  def to_rb
+    return nil if self == JS::Null
+    case self.typeof
+    when "number"
+      # TODO: HTTP Codes end up as 200.0, check if it could be integer?
+      # In JS all numbers are floating point.
+      # Is there float.to_js? Create tests for number conversion. Check Ruby JSON parser.
+      self.to_f
+    when "string"
+      self.to_s
+    when "boolean"
+      self == JS::True
+    when "symbol" # TODO: check if this works with assingment
+      self.to_sym
+    when "bigint"
+      self.to_i
+    when "object"
+      #if self.call(:instanceof, JS.global[:Date]) and not JS.global.isNaN(self)
+      #  self.toISOString()
+      #elsif JS.global[:Array].isArray(self)
+      self.isJSArray ? self.to_a : self
+    else
+      self
+    end
+    #return self.static_to_rb(self)
+
+    #case self[sym]
+
+    # TODO convert js to numbers integers to
+    # and change this from the method_missing
+    # make sure to run it in to _a
+
+    # JS::True
+    # Date to Ruby Date
+  end
+
+  # Support self == true instead of self == JS:True
+  alias_method :orig_eq, :==
+  def ==(other)
+    if other.equal? true
+      return orig_eq(JS::True)
+    elsif other.equal? false
+      return orig_eq(JS::False)
+    elsif other.equal? nil
+      return orig_eq(JS::Null) || orig_eq(JS::Undefined)
+    end
+
+    orig_eq(other)
+  end
+
+  def isJSArray()
+    JS.global[:Array].isArray(self) == JS::True
+  end
+
+  def self.static_to_rb(object)
+    return nil if self[sym] == JS::Null
+  end
+
+  def each(&block)
+    if block_given?
+      self.isJSArray ? self.to_a.each(&block) : Array.new(__props).each(&block)
```

Comment:
Hmm, I feel iterating on object properties is a little bit tricky. Do you have any use case for this behavior? I prefer simply delegating to `self.to_a.each`